consumer.go

```diff
@@ -0,0 +1,286 @@
+)
+
+type Consumer struct {
+	Name               string
+	ConsumerGroup      string
+	PullIntervalMillis int
+	BatchSize          int
+	MaxAckTimeMillis   int
+	MaxMsgDeliveries   int
+	BatchMaxTimeToWait time.Duration
+	conn               *Conn
+	stationName        string
+	subscription       *nats.Subscription
+	firstFetch         bool
+	consumeActive      bool
+	puller             chan *Msg
+	pullerQuit         chan struct{}
+}
+
+type Msg struct {
+	msg *nats.Msg
+}
+
+func (m *Msg) Data() []byte {
+	return m.msg.Data
+}
+
+func (m *Msg) Ack() error {
+	return m.msg.Ack()
+}
+
+type createConsumerReq struct {
+	Name             string `json:"name"`
+	StationName      string `json:"station_name"`
+	ConnectionId     string `json:"connection_id"`
+	ConsumerType     string `json:"consumer_type"`
+	ConsumerGroup    string `json:"consumers_group"`
+	MaxAckTimeMillis int    `json:"max_ack_time_ms"`
+	MaxMsgDeliveries int    `json:"max_msg_deliveries"`
+}
+
+type removeConsumerReq struct {
+	Name        string `json:"name"`
+	StationName string `json:"station_name"`
+}
+
+type ConsumerOpts struct {
+	Name                     string
+	StationName              string
+	ConsumerGroup            string
+	PullIntervalMillis       int
+	BatchSize                int
+	BatchMaxTimeToWaitMillis int
+	MaxAckTimeMillis         int
+	MaxMsgDeliveries         int
+}
+
+func GetDefaultConsumerOptions() ConsumerOpts {
+	return ConsumerOpts{
+		PullIntervalMillis:       1000,
+		BatchSize:                10,
+		BatchMaxTimeToWaitMillis: 5000,
+		MaxAckTimeMillis:         30000,
+		MaxMsgDeliveries:         10,
+	}
+}
+
+type ConsumerOpt func(*ConsumerOpts) error
+
+func (c *Conn) CreateConsumer(stationName, consumerName string, opts ...ConsumerOpt) (*Consumer, error) {
+	defaultOpts := GetDefaultConsumerOptions()
+
+	defaultOpts.Name = consumerName
+	defaultOpts.StationName = stationName
+	defaultOpts.ConsumerGroup = consumerName
+
+	for _, opt := range opts {
+		if opt != nil {
+			if err := opt(&defaultOpts); err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	return defaultOpts.CreateConsumer(c)
+}
+
+func (opts *ConsumerOpts) CreateConsumer(c *Conn) (*Consumer, error) {
+	consumer := Consumer{Name: opts.Name,
+		ConsumerGroup:      opts.ConsumerGroup,
+		PullIntervalMillis: opts.PullIntervalMillis,
+		BatchSize:          opts.BatchSize,
+		MaxAckTimeMillis:   opts.MaxAckTimeMillis,
+		MaxMsgDeliveries:   opts.MaxMsgDeliveries,
+		BatchMaxTimeToWait: time.Duration(opts.BatchMaxTimeToWaitMillis) * time.Millisecond,
+		conn:               c,
+		stationName:        opts.StationName}
+
+	err := c.create(&consumer)
+	if err != nil {
+		return nil, err
+	}
+
+	consumer.firstFetch = true
+	consumer.puller = make(chan *Msg, consumer.BatchSize)
+	consumer.pullerQuit = make(chan struct{}, 1)
+
+	ackWait := time.Duration(consumer.MaxAckTimeMillis) * time.Millisecond
+	subj := consumer.stationName + ".final"
+	durableName := consumer.ConsumerGroup
+	if durableName == "" {
```

Comment:
since we decided to initialize the group name to the consumer name this if statement is unnecessary